stackrox/stackrox
# ROX-9843: Add expose monitoring flag to operator

## Description

With this PR we are adding an option to operator configuration to enable monitoring for Central. This is required for proper integration with the observability operator installed together with ACS Fleetshard. Related PR on ASC Fleet manager: https://github.com/stackrox/acs-fleet-manager/pull/112

The primary function of the operator is to provide values for the Helm template, and everything will be installed/updated via helm. Our helm template already provides `central.exposeMonitoring` boolean value, which is used to add the monitoring port to the central pod: https://github.com/stackrox/stackrox/blob/210afd36677ab3e6b66ca6dc512d3d075b2450ff/image/templates/helm/stackrox-central/templates/_central_endpoints.tpl#L47-L50  and also to deploy relevant network policy for it https://github.com/stackrox/stackrox/blob/210afd36677ab3e6b66ca6dc512d3d075b2450ff/image/templates/helm/stackrox-central/templates/01-central-10-networkpolicy.yaml#L21-L26 This PR provides a possibility to set that value (flag) via operator CRD.

**Made changes**

- new flag added to the operator configuration
- tests when the flag is enabled and disabled

**Open question**

- Should we add information about the new flag to Changelog?

## Checklist
- ~[ ] Determined and documented upgrade steps~ - no need for upgrade steps
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~ - also not relevant

## Testing Performed 

1. Letting CI run the KUTTL tests
2. Start OS4 cluster with `infractl`
3. Deploy operator to the new cluster (using images built by CI pipeline). i.e.
```
VERSION=3.70.0-414-gd49bd107a4 make deploy
```
**NOTE** It can happen that we need to adjust pull image secrets if the cluster is not able to pull images
4. Create a Central CRD resource (without monitoring first). First, we need a new project and then create CRD inside that project. An example can be taken from the test: `operator/tests/central/basic/10-central-cr.yaml`
5. Check that everything is deployed properly
6. Set `exposeMonitoring` flag on the previously created Central resource
7. Validate that the central pod has an additional port and that network policy is created

**Test full data plane observability flow**
1. Deploy observability helm provided by PR: https://github.com/stackrox/acs-fleet-manager/pull/112
2. Validate that the observability operator is installed properly and that metrics are collected. For that, we need to open OS4 UI and navigate to Routes of the project where the observability stack is deployed
3. in Prometheus UI - `/graph`, check if `rox_*` metrics are visible
4. checking other parts like Grafana, Observatorium, and Alert Manager are not relevant for this PR. It's important that Prometheus is able to collect metrics, everything else relies on that.
